common/src/autogluon/common/utils/deprecated.py

```diff
@@ -1,26 +1,230 @@
-import logging
-import warnings
 import functools
-from typing import Optional
+import inspect
+import warnings
+from typing import Any, Dict, Optional
+
+from packaging import version
 
+from ..version import __version__
 
-logger = logging.getLogger(__name__)
 
+def _deprecation_warning(
+    old: str,
+    new: Optional[str] = None,
+    custom_warning_msg: Optional[str] = None,
+    version_to_remove: Optional[str] = None,
+    error: bool = False,
+):
+    msg = custom_warning_msg
+    if msg is None:
+        msg = f"`{old}` has been deprecated. "
+        msg += f"And will be removed in version {version_to_remove} " if version_to_remove is not None else ""
+        msg += f"Please use `{new}` instead" if new is not None else ""
 
-def deprecated(message: str = "", version_removed: Optional[str] = None, *args, **kwargs):
-    def deprecated_decorator(func):
-        warning_message = (
-            f"Calling deprecated function {func.__name__}. This function is deprecated "
-            + "and will be removed in "
-            + (f"v{version_removed}. " if version_removed else "a future release. ")
-            + message
+    if error:
+        raise ValueError(msg)
+    else:
+        warnings.warn(
+            f"Deprecation Warning: {msg}. This will raise an error in the future!",
+            category=DeprecationWarning,
+            stacklevel=2,
         )
 
-        @functools.wraps(func)
-        def wrapped_func(*args, **kwargs):
-            warnings.warn(message=warning_message, category=DeprecationWarning, stacklevel=2)
-            return func(*args, **kwargs)
 
-        return wrapped_func
+def Deprecated(
```

Comment:
Is there a convention you are following for why it is capitalized?

Reply:
I've seen ray warpping it's api call with Capitalized decorator, so trying to mimic here. I feel that makes it pretty noticeable

Reply:
sounds good